### DISCUSSION TITLE
Inference/ prediction in Tutorial.iypnb

### DISCUSSION THREAD

**Discussion by Atmoboran** [terrastackai/terratorch#492] [Q&A]
Hello,

so I have been playing around with examples/notebooks/Tutorial.ipynb , which I find very helpful to understand the whole workflow of loading data and training the model!
What is missing from my point of view is the inference step. How would one infer the model inside a jupyter notebook (not just trainer.test)? I know that if one doesn't train with more data than given in the example the inference will not produce any reasonable results, but it would still be nice to see how it would work if trained longer and on more data.

Best regards,
Boran

**Comment by Joao-L-S-Almeida** [comment]
Hi, @Atmoboran 
Good point. I'll add it to examples/notebooks/Tutorial.ipynb. Maybe some notebooks showing simple inference for others large (and well trained) models would also be helpful.

**Reply by Atmoboran** [reply]
So how would I do the inference? I have tried some things, but the predictions look random.
I have trained the burnscars model like in the tutorial, but with 10 epochs and 200 training images with the encoder and decoder NOT frozen (Is that enough?). Then I'm trying to do inference like in the following:

# Load weights saved after training
model_1 = task.model 
end_model_weight_path = "model_weights.pth"
model_1.load_state_dict(torch.load(end_model_weight_path))
model_1.eval()  # Set to inference mode

# Select a batch of test images
batch = next(iter(datamodule.test_dataloader()))
test_images = batch["image"]  # Extract images from batch
test_labels = batch["mask"]  # Ground truth masks (option
# Run inference
with torch.no_grad():
    predictions = model(test_images)

# Extract logits from predictions
logits = predictions.output  # Shape: (12, 2, 512, 512)

# Convert logits to predicted class masks
predicted_masks = logits.argmax(dim=1).cpu().numpy()  # Shape: (12, 512, 512)

**Comment by romeokienzler** [comment]
Hi @Atmoboran 

here an inference example taken from https://github.com/IBM/terratorch/blob/main/examples/notebooks/WxCTutorialDownscaling.ipynb
`

dm = Merra2DownscaleNonGeoDataModule(
    time_range=('2020-01-01T00:00:00', '2020-01-01T23:59:59'),
    data_path_surface = config.data.data_path_surface,
    data_path_vertical = config.data.data_path_vertical,
    climatology_path_surface = config.data.climatology_path_surface,
    climatology_path_vertical = config.data.climatology_path_vertical,
    input_surface_vars = config.data.input_surface_vars,
    input_static_surface_vars = config.data.input_static_surface_vars,
    input_vertical_vars = config.data.input_vertical_vars,
    input_levels = config.data.input_levels,
    n_input_timestamps = config.data.n_input_timestamps,
    output_vars=config.data.output_vars,
    transforms=_get_transforms(config),
)
dm.setup('predict')

trainer = Trainer(
    max_epochs=1,
    limit_predict_batches=1,
)
results = trainer.predict(model=task, datamodule=dm)

`
